File: src/bmsx_cpp/machine/runtime/frame/loop.cpp
#include "machine/runtime/frame/loop.h"
#include "machine/runtime/cart_boot.h"
#include "machine/runtime/cpu_executor.h"
#include "machine/runtime/runtime.h"

namespace bmsx {
void FrameLoopState::reset() {
	frameDeltaMs = 0.0;
	currentTimeSeconds = 0.0;
}

void FrameLoopState::resetFrameState(Runtime& runtime) {
	frameActive = false;
	runtime.vblank.abandonTick();
	runtime.machine.inputController.cancelSampleArm();
	frameState = FrameState{};
	runtime.machine.cpu.clearHaltUntilIrq();
	runtime.frameScheduler.reset();
	reset();
	runtime.screen.reset();
	runtime.frameScheduler.resetTickTelemetry();
	runtime.vblank.reset(runtime);
}

void FrameLoopState::beginFrameState(Runtime& runtime) {
	frameActive = true;
	runtime.vblank.beginTick();
	frameState = FrameState{};
	frameState.cycleBudgetRemaining = runtime.timing.cycleBudgetPerFrame;
	frameState.cycleBudgetGranted = runtime.timing.cycleBudgetPerFrame;
	frameState.cycleCarryGranted = 0;
	frameDeltaMs = runtime.timing.frameDurationMs;
	runtime.machine.vdp.beginFrame();
}

void FrameLoopState::abandonFrameState(Runtime& runtime) {
	frameActive = false;
	runtime.vblank.abandonTick();
}

void FrameLoopState::finalizeUpdateSlice(Runtime& runtime) {
	if (runtime.m_pendingCall == Runtime::PendingCall::Entry && !runtime.vblank.tickCompleted()) {
		return;
	}
	abandonFrameState(runtime);
}

void FrameLoopState::runUpdatePhase(Runtime& runtime) {
	try {
		while (true) {
			if (runtime.machine.cpu.isHaltedUntilIrq() && runtime.cpuExecution.runHaltedUntilIrq(runtime, frameState)) {
				return;
			}
			if (runtime.m_pendingCall != Runtime::PendingCall::Entry) {
				return;
			}
			const RunResult result = runtime.cpuExecution.runWithBudget(runtime, frameState);
			if (runtime.machine.cpu.isHaltedUntilIrq()) {
				if (runtime.cpuExecution.runHaltedUntilIrq(runtime, frameState)) {
					return;
				}
				continue;
			}
			if (result == RunResult::Halted) {
				runtime.m_pendingCall = Runtime::PendingCall::None;
			}
			return;
		}
	} catch (...) {
		frameState.luaFaulted = true;
		runtime.machine.cpu.clearHaltUntilIrq();
		runtime.m_pendingCall = Runtime::PendingCall::None;
		throw;
	}
}

bool FrameLoopState::tickUpdate(Runtime& runtime) {
	using PendingCall = Runtime::PendingCall;
	if (!runtime.m_luaInitialized || !runtime.m_tickEnabled || runtime.m_runtimeFailed) {
		return false;
	}

	if (runtime.cartBoot.processPending()) {
		return true;
	}

	const bool previousFrameActive = frameActive;
	const int previousRemaining = previousFrameActive ? frameState.cycleBudgetRemaining : -1;
	const bool previousPending = runtime.m_pendingCall == PendingCall::Entry;
	const i64 previousSequence = runtime.frameScheduler.lastTickSequence;
	const bool startedFrame = !frameActive;
	if (frameActive) {
		if (frameState.cycleBudgetRemaining <= 0 && !runtime.frameScheduler.refillFrameBudget(runtime, frameState)) {
			return false;
		}
	} else {
		if (!runtime.frameScheduler.startScheduledFrame(runtime)) {
			return false;
		}
	}

	if (runtime.m_pendingCall == PendingCall::Entry) {
		runUpdatePhase(runtime);
	}

	if (startedFrame) {
		runtime.m_debugUpdateCountTotal += 1;
	}

	frameState.updateExecuted = runtime.m_pendingCall != PendingCall::Entry;
	finalizeUpdateSlice(runtime);
	const bool nextFrameActive = frameActive;
	if (nextFrameActive != previousFrameActive) {
		return true;
	}
	if (nextFrameActive && frameState.cycleBudgetRemaining != previousRemaining) {
		return true;
	}
	if ((runtime.m_pendingCall == PendingCall::Entry) != previousPending) {
		return true;
	}
	return runtime.frameScheduler.lastTickSequence != previousSequence;
}
} // namespace bmsx

File: packages/bmsx-console/src/machine/runtime/frame/loop.ts
import { FrameState, Runtime } from '../runtime';

export class FrameLoopState {
	public currentTimeMs = 0;
	public frameDeltaMs = 0;
	public readonly frameState: FrameState = {
		haltGame: false,
		updateExecuted: false,
		luaFaulted: false,
		cycleBudgetRemaining: 0,
		cycleBudgetGranted: 0,
		cycleCarryGranted: 0,
		activeCpuUsedCycles: 0,
	};
	public frameActive = false;

	constructor(private readonly runtime: Runtime) {
	}

	public reset(): void {
		this.currentTimeMs = 0;
		this.frameDeltaMs = 0;
	}

	public resetFrameState(): void {
		const runtime = this.runtime;
		this.abandonFrameState();
		const state = this.frameState;
		state.haltGame = false;
		state.updateExecuted = false;
		state.luaFaulted = false;
		state.cycleBudgetRemaining = 0;
		state.cycleBudgetGranted = 0;
		state.cycleCarryGranted = 0;
		state.activeCpuUsedCycles = 0;
		runtime.machine.cpu.clearHaltUntilIrq();
		runtime.frameScheduler.reset();
		this.reset();
		runtime.screen.reset();
		runtime.frameScheduler.resetTickTelemetry();
	}

	public beginFrameState(): FrameState {
		if (this.frameActive) {
			throw new Error('attempted to begin a new frame while another frame is active.');
		}
		const runtime = this.runtime;
		this.frameDeltaMs = runtime.timing.frameDurationMs;
		const budget = runtime.timing.cycleBudgetPerFrame;
		const state = this.frameState;
		state.haltGame = runtime.debuggerPaused;
		state.updateExecuted = false;
		state.luaFaulted = runtime.luaRuntimeFailed;
		state.cycleBudgetRemaining = budget;
		state.cycleBudgetGranted = budget;
		state.cycleCarryGranted = 0;
		state.activeCpuUsedCycles = 0;
		runtime.machine.vdp.beginFrame();
		runtime.vblank.beginTick();
		this.frameActive = true;
		return state;
	}

	public tickUpdate(): boolean {
		const runtime = this.runtime;
		if (!runtime.tickEnabled) {
			return false;
		}
		if (runtime.cartBoot.processPending()) {
			return true;
		}
		if (runtime.executionOverlayActive) {
			if (this.frameActive) {
				this.abandonFrameState();
				return true;
			}
			return false;
		}
		const previousFrameActive = this.frameActive;
		const previousRemaining = previousFrameActive ? this.frameState.cycleBudgetRemaining : -1;
		const frameScheduler = runtime.frameScheduler;
		const previousPendingEntry = runtime.pendingCall === 'entry';
		const previousSequence = frameScheduler.lastTickSequence;
		if (!this.frameActive) {
			if (!frameScheduler.startScheduledFrame()) {
				return false;
			}
		} else if (this.frameState.cycleBudgetRemaining <= 0) {
			if (!frameScheduler.refillFrameBudget(this.frameState)) {
				return false;
			}
		}
		const haltedUntilIrq = this.runActiveFrameState();
		if (haltedUntilIrq) {
			return false;
		}
		const nextFrameActive = this.frameActive;
		if (nextFrameActive !== previousFrameActive) {
			return true;
		}
		if (nextFrameActive && this.frameState.cycleBudgetRemaining !== previousRemaining) {
			return true;
		}
		const nextPendingCall = runtime.pendingCall;
		if ((nextPendingCall === 'entry') !== previousPendingEntry) {
			return true;
		}
		const nextSequence = frameScheduler.lastTickSequence;
		return nextSequence !== previousSequence;
	}

	public abandonFrameState(): void {
		this.frameActive = false;
		const runtime = this.runtime;
		runtime.vblank.abandonTick();
	}

	private runActiveFrameState(): boolean {
		const runtime = this.runtime;
		const state = this.frameState;
		if (runtime.pendingCall === 'entry') {
			const haltedUntilIrq = this.runUpdatePhase();
			state.updateExecuted = runtime.pendingCall !== 'entry';
			this.finalizeUpdateSlice();
			return haltedUntilIrq;
		}
		this.finalizeUpdateSlice();
		return false;
	}

	private finalizeUpdateSlice(): void {
		const runtime = this.runtime;
		if (runtime.pendingCall === 'entry' && !runtime.vblank.tickCompleted) {
			return;
		}
		this.abandonFrameState();
	}

	private runUpdatePhase(): boolean {
		const runtime = this.runtime;
		const state = this.frameState;
		const cpu = runtime.machine.cpu;
		const cpuExecution = runtime.cpuExecution;
		if (!runtime.cartEntryAvailable) {
			return false;
		}
		if (!runtime.luaGate.ready) {
			return false;
		}
		if (state.luaFaulted || runtime.luaRuntimeFailed) {
			state.luaFaulted = true;
			return false;
		}
		if (state.haltGame) {
			return false;
		}
		try {
			while (true) {
				if (cpu.isHaltedUntilIrq()) {
					const tickCompleted = cpuExecution.runHaltedUntilIrq(state);
					if (tickCompleted || cpu.isHaltedUntilIrq()) {
						return true;
					}
					continue;
				}
				if (runtime.pendingCall !== 'entry') {
					return false;
				}
				runtime.cpuExecution.runWithBudget(state);
				if (cpu.isHaltedUntilIrq()) {
					return true;
				}
				return false;
			}
		} catch (error) {
			state.luaFaulted = true;
			cpu.clearHaltUntilIrq();
			runtime.pendingCall = null;
			throw error;
		}
	}
}
